realm/realm-library/src/main/cpp/io_realm_internal_Table.cpp

```diff
@@ -232,12 +243,22 @@ JNIEXPORT void JNICALL Java_io_realm_internal_Table_nativeConvertColumnToNullabl
             j++;
         }
 
+        // Search index has too be added first since if it is a PK field, add_xxx_unique will check it.
+        if (table->has_search_index(column_index + 1)) {
+            table->add_search_index(column_index);
+        }
+
         for (size_t i = 0; i < table->size(); ++i) {
             switch (column_type) {
                 case type_String: {
                     // Payload copy is needed
                     StringData sd(table->get_string(column_index + 1, i));
-                    table->set_string(column_index, i, sd);
+                    if (is_primary_key) {
+                        table->set_string_unique(column_index, i, sd);
```

Comment:
This looks kind of a critical bug fix, since it previously could have caused rows to merge incorrectly?